### PR TITLE
Add a new menu for the admin in order to check the quality of the import

### PR DESCRIPTION
**Pull request by loikki**
In Sponsorship->Configuration->Test Imports admin can do an import
(with an additionnal csv) that will compare the values obtained and
the value in the csv.

### CODE REVIEW

**Comment by loikki** [comment]
No need to review yet, I still want to add something to the preview

File: sbc_compassion/wizards/import_letters_test.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import base64
+
+from openerp import models, fields, _, api
+from ..tools import import_letter_functions as func
+
+
+class TestImportLetters(models.TransientModel):
+    """ Class testing the letters import by highlighting the keypoints of
+    the import (image used, keypoints found for the templates, showing only
+    the line not well detected, ...)
```
Comment (ecino) [inline sbc_compassion/wizards/import_letters_test.py]: Avoid ... in comments

File: sbc_compassion/wizards/import_letters_test.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    """
+    _inherit = 'import.letters.history'
+    _name = 'test.import.letters.history'
+    _description = _("""Easier way to test the import""")
+    csv_file = fields.Many2many(
```
Comment (ecino) [inline sbc_compassion/wizards/import_letters_test.py]: Can it have many csv attached ? Please respect the standard convention for field names :
Many2one should end by `_id` and Many2many should end with `_ids`

File: sbc_compassion/wizards/import_letters_test.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class TestImportLetters(models.TransientModel):
+    """ Class testing the letters import by highlighting the keypoints of
+    the import (image used, keypoints found for the templates, showing only
+    the line not well detected, ...)
+    """
+    _inherit = 'import.letters.history'
+    _name = 'test.import.letters.history'
+    _description = _("""Easier way to test the import""")
```
Comment (ecino) [inline sbc_compassion/wizards/import_letters_test.py]: This is not a good description. Either remove it or be more explanatory.

File: sbc_compassion/tools/import_letter_functions.py [hunk 4]
```diff
@@ -21,13 +21,117 @@
...
+        csv_string = base64.b64decode(csv_.datas).splitlines()
+        csv_data += read_csv(csv_string)
+    # check the letters
+    values = {}
+    test = True
+    for csv_line in csv_data:
+        if csv_line['name'] == doc_name:
+            values = csv_line
+            break
+    if values == {}:
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: `if not values` is better for testing emptiness of dictionary.

File: sbc_compassion/tools/import_letter_functions.py [hunk 5]
```diff
@@ -21,13 +21,117 @@
...
+        raise Warning(
+            _("File '{}' not found in the csv files".format(doc_name)))
+    # first test: template
+    template_id = env.ref(
+        'sbc_compassion.'+values['template_id'])
+    tpl_detected = env['sponsorship.correspondence.template'].search(
+        [('id', '=', line_vals['template_id'])])
+    temp_test = template_id == tpl_detected
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: I think there is no need to have `temp_test` and `test` variables. You can simply do :
```
if not template_id == tpl_detected:
    error = error + 'template,\n'
...
```

File: sbc_compassion/tools/import_letter_functions.py [hunk 6]
```diff
@@ -21,13 +21,117 @@
...
+    # partner
+    partner_temp = env['res.partner'].search([
+        ('id', '=', line_vals['partner_id'])])
+    # child
+    child_temp = env['compassion.child'].search([
+        ('id', '=', line_vals['child_id'])])
+    pcodega = partner_temp.ref
+    ccode = child_temp.code
+    temp_test = True
+    if values['barcode'] == '' and not (pcodega and ccode):
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Can you explain the following test and what do you capture in `temp_test` ? I have the feeling this can be greatly simplified (until line 96).

File: sbc_compassion/tools/import_letter_functions.py [hunk 7]
```diff
@@ -21,13 +21,117 @@
...
+        temp_test = temp_test and lang in values['lang']
+    line_nber_lang = len(line_vals['test_letter_languages_id'].split(','))
+    csv_nber_lang = len(values['lang'].split(','))
+    temp_test = temp_test and line_nber_lang == csv_nber_lang
+
+    test = temp_test and test
+    if not temp_test:
+        error = error + 'lang,\n'
+
+    return test, error[:-2]
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Add a comment explaining you remove end of line in error string. You could avoid to return `test` variable and instead check outside of the func:

```
if error:
    .... (means an error was detected and error string is not empty)
else:
    .... (no error detected)
```

File: sbc_compassion/tools/import_letter_functions.py [hunk 8]
```diff
@@ -21,13 +21,117 @@
...
+    # child
+    child_temp = env['compassion.child'].search([
+        ('id', '=', line_vals['child_id'])])
+    pcodega = partner_temp.ref
+    ccode = child_temp.code
+    temp_test = True
+    if values['barcode'] == '' and not (pcodega and ccode):
+        temp_test = True
+    # xor (if both pcodega and ccode are False => barcode not detected)
+    elif bool(pcodega) ^ bool(ccode):
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Don't understand what is your test

File: sbc_compassion/tools/import_letter_functions.py [hunk 9]
```diff
@@ -21,13 +21,117 @@
...
+    temp_test = temp_test and line_nber_lang == csv_nber_lang
+
+    test = temp_test and test
+    if not temp_test:
+        error = error + 'lang,\n'
+
+    return test, error[:-2]
+
+
+def update_text(text, test):
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Not a good method name, it is too generic.

File: sbc_compassion/tools/import_letter_functions.py [hunk 10]
```diff
@@ -21,13 +21,117 @@
...
 from wand.image import Image
 
+from openerp import _
+
 
 ##########################################################################
 #                           GENERAL METHODS                              #
 ##########################################################################
+def testline(env, line_vals, csv_file, doc_name):
+    """ Test each line and verify the data given by the csv.
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Do you document anywhere what the csv file looks like ? Is it the same as the one used for the tests ? There should be at least somwhere inside the module a explanation about it (if there already is one, you can point to it inside this docstring).

File: sbc_compassion/tools/import_letter_functions.py [hunk 11]
```diff
@@ -21,13 +21,117 @@
...
+    If a line does not correspond to the csv, this function erases it.
+    :param env: Odoo environment
+    :param dict line_vals: Dictionnary that will be used to create line
+    :param One2many('test.import.letter.line') csv_file: CSV used to check
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: This doesn't make sense to me : a param cannot be a One2many. One2many is a field, It can be a Recordset of 'test.import.letter.line' objects. But is it a recordset or a CSV file ?

File: sbc_compassion/wizards/import_letters_test.py (hunk 2 above)
Comment (loikki) [inline sbc_compassion/wizards/import_letters_test.py]: And One2many with _ids?

File: sbc_compassion/tools/import_letter_functions.py [hunk 12]
```diff
@@ -21,13 +21,135 @@
...
+    temp_test = temp_test and line_nber_lang == csv_nber_lang
+
+    test = temp_test and test
+    if not temp_test:
+        error = error + 'lang,\n'
+
+    return test, error[:-2]
+
+
+def update_text(test_import):
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Not a good method name, it is too generic.

File: sbc_compassion/tools/import_letter_functions.py [hunk 13]
```diff
@@ -66,13 +188,18 @@ def analyze_attachment(env, file_, filename, force_template, test=False):
...
     # loop over all the patterns in the pattern directory
-    line_vals, pattern_center = _find_template(env, img, line_vals)
+    line_vals, pattern_center = _find_template(env, img, line_vals, test)
     if line_vals['template_id'] != env.ref('sbc_compassion.default_template'):
-        line_vals = _find_languages(env, img, pattern_center, line_vals)
+        line_vals = _find_languages(env, img, pattern_center, line_vals, test)
     else:
         line_vals.update({
-            'letter_language_id': False
+            'letter_language_id': None
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Usually, False reprensents an empty relation (not None).

File: sbc_compassion/tools/import_letter_functions.py (hunk 5 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: I need to return test, therefore I am avoiding to test twice the equality by writing it inside a variable (and it decreases the size of a line)

File: sbc_compassion/tools/import_letter_functions.py [hunk 14]
```diff
@@ -242,12 +383,13 @@ def _save_img(MULTIPAGE, file_, img):
...
 
 
-def _find_template(env, img, line_vals):
+def _find_template(env, img, line_vals, test):
     """
     Use pattern recognition to detect which template corresponds to img.
     :param env env: Odoo variable env
     :param img: Image to analyze
     :param dict line_vals: Dictonnary containing the data for a line
+    :param bool test: Save the image of all the template detections
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: This is really a bad choice for variable name if it does what it says in this comment.

File: sbc_compassion/tools/import_letter_functions.py [hunk 15]
```diff
@@ -302,13 +454,26 @@ def _find_languages(env, img, pattern_center, line_vals):
     :param pattern_center: Center position of detected pattern
     :param dict line_vals: Dictonnary containing the data for a line
     (and the template)
+    :param bool test: Save the image of all the subimage
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Again bad name or bad comment. Comment should be `tells if the import is done in test mode` or variable name should be `save_images`

File: sbc_compassion/tools/import_letter_functions.py (hunk 6 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: I have found a small mistake in it, I am rewriting it

File: sbc_compassion/wizards/import_letters_test.py (hunk 2 above)
Comment (ecino) [inline sbc_compassion/wizards/import_letters_test.py]: Yes

File: sbc_compassion/tools/import_letter_functions.py (hunk 5 above)
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: I think you don't need test. See my comment at the return line.

File: sbc_compassion/tools/patternrecognition.py [hunk 16]
```diff
@@ -120,7 +120,7 @@ def findMatches(des1, des2, test=0.8):
     :rtype: list[Keypoints]
     """
     bf = cv2.BFMatcher()
-    matches = bf.knnMatch(des2, des1, k=2)
+    matches = bf.knnMatch(des1, des2, k=2)
```
Comment (ecino) [inline sbc_compassion/tools/patternrecognition.py]: Why this swap, can you explain the change?

File: sbc_compassion/tools/patternrecognition.py [hunk 17]
```diff
@@ -82,7 +82,7 @@ def patternRecognition(image, pattern, crop_area=None,
         return kp1, kp2, good
     if len(good) >= threshold:
         # put in a np.array the position of the image's keypoints
-        keypoints = np.array([kp1[i[0].trainIdx].pt for i in good])
+        keypoints = np.array([kp1[i[0].queryIdx].pt for i in good])
```
Comment (ecino) [inline sbc_compassion/tools/patternrecognition.py]: Can you explain this change ?

File: sbc_compassion/tools/patternrecognition.py [hunk 18]
```diff
@@ -167,7 +167,7 @@ def keyPointCenter(keypoints):
         return center / N
 
 
-def find_template(img, templates, threshold=0.8):
+def find_template(img, templates, test=False, threshold=0.8):
```
Comment (ecino) [inline sbc_compassion/tools/patternrecognition.py]: you forgot to update the method docstring.

File: sbc_compassion/tools/patternrecognition.py [hunk 19]
```diff
@@ -197,5 +215,7 @@ def find_template(img, templates, threshold=0.8):
                 len(tmp_key))/float(template.nber_keypoints)
             key_img = tmp_key
             matching_template = template
-
-    return matching_template, keyPointCenter(key_img)
+    if test:
```
Comment (ecino) [inline sbc_compassion/tools/patternrecognition.py]: Why not always return test_img (can be none if not called with test) ?

File: sbc_compassion/tools/import_letter_functions.py (hunk 6 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: Done

File: sbc_compassion/wizards/import_letters_test.py (hunk 2 above)
Comment (loikki) [inline sbc_compassion/wizards/import_letters_test.py]: I forgot to answer to the question. Yes it can (it is easy to deal with it in the code and the user will not have to create a new csv every time he wants to mix some test files together)

File: sbc_compassion/tools/import_letter_functions.py (hunk 10 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: I need to check with rz-comp

File: sbc_compassion/tools/patternrecognition.py (hunk 16 above)
Comment (loikki) [inline sbc_compassion/tools/patternrecognition.py]: I had some trouble with the images for the test and usually this order is given (and seems more logical). I do not know exactly the difference, but I think I have done it wrong before this modification

File: sbc_compassion/tools/patternrecognition.py (hunk 17 above)
Comment (loikki) [inline sbc_compassion/tools/patternrecognition.py]: Same as the comment above.
As I have done a switch between des1 and des2, I need to take the other Id (query is for one image and train the other one).

File: sbc_compassion/tools/patternrecognition.py (hunk 18 above)
Comment (loikki) [inline sbc_compassion/tools/patternrecognition.py]: Done

File: sbc_compassion/tools/import_letter_functions.py [hunk 20]
```diff
@@ -21,13 +21,134 @@
...
 ##########################################################################
+def testline(env, line_vals, csv_file_ids, doc_name):
+    """ Test each line and verify the data given by the csv.
+    If a line does not correspond to the csv, this function erases it.
+    A csv file need to be written in the same way than
+    sbc_compassion/tests/testdata/import/travis_file.csv
+
+    :param env: Odoo environment
+    :param dict line_vals: Dictionnary that will be used to create line
+    :param Recordset('test.import.letter.line') csv_file_ids: CSV used to check
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: In fact after checking, csv_file_ids are of type `ir.attachement`, not `test.import.letter.line`

File: sbc_compassion/tools/import_letter_functions.py [hunk 21]
```diff
@@ -21,13 +21,134 @@
...
+    # check the letters
+    values = {}
+    for csv_line in csv_data:
+        if csv_line['name'] == doc_name:
+            values = csv_line
+            break
+    if not values:
+        raise Warning(
+            _("File '{}' not found in the csv files".format(doc_name)))
+    # first test: template
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: `Test that the template is the same as given in the csv`

File: sbc_compassion/tools/import_letter_functions.py [hunk 22]
```diff
@@ -21,13 +21,134 @@
...
+    # str that will be returned
+    error = ''
+    csv_data = []
+    # creates a list containing all the csv data
+    for csv_ in csv_file_ids:
+        if os.path.splitext(csv_.name)[1] != '.csv':
+            raise Warning(_("File '{}' is not a csv file".format(csv_.name)))
+        csv_string = base64.b64decode(csv_.datas).splitlines()
+        csv_data += read_csv(csv_string)
+    # check the letters
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: `Check that the letter is present in the csv file`

File: sbc_compassion/tools/import_letter_functions.py [hunk 23]
```diff
@@ -21,13 +21,134 @@
...
+    values = {}
+    for csv_line in csv_data:
+        if csv_line['name'] == doc_name:
+            values = csv_line
+            break
+    if not values:
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Instead of this, you can use a `for.. else` loop (check the syntax)

File: sbc_compassion/tools/import_letter_functions.py [hunk 24]
```diff
@@ -21,13 +21,134 @@
...
+        raise Warning(
+            _("File '{}' not found in the csv files".format(doc_name)))
+    # first test: template
+    template_id = env.ref(
+        'sbc_compassion.'+values['template_id'])
+    tpl_detected = env['sponsorship.correspondence.template'].search(
+        [('id', '=', line_vals['template_id'])])
+    # test if the template detected is the same as the expected one
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: You can remove this comment if the previous is updated.

File: sbc_compassion/tools/import_letter_functions.py [hunk 25]
```diff
@@ -21,13 +21,134 @@
...
+    values = {}
+    for csv_line in csv_data:
+        if csv_line['name'] == doc_name:
+            values = csv_line
+            break
+    if not values:
+        raise Warning(
+            _("File '{}' not found in the csv files".format(doc_name)))
+    # first test: template
+    template_id = env.ref(
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: `template_csv`

Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: template_csv_id?

File: sbc_compassion/tools/import_letter_functions.py [hunk 26]
```diff
@@ -21,13 +21,134 @@
...
+        'sbc_compassion.'+values['template_id'])
+    tpl_detected = env['sponsorship.correspondence.template'].search(
+        [('id', '=', line_vals['template_id'])])
+    # test if the template detected is the same as the expected one
+    if template_id != tpl_detected:
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: I think it is safer to compare their ids instead of the entire objects:
`template_csv.id != tpl_detected.id`

File: sbc_compassion/tools/import_letter_functions.py [hunk 27]
```diff
@@ -21,13 +21,134 @@
...
+        error = error + 'template,\n'
+
+    # second: qr codexb
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Please write sentences in the comments.
`Test that the barcode detection matches the result given in csv file.`

File: sbc_compassion/tools/import_letter_functions.py (hunk 25 above)
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Sorry for those rules but `_id` suffix is for field names only (because they end up to be rows in the database storing integers). However, for only variable names, it makes less sense to suffix with `_id` as the variable will be a record and not an integer value.

File: sbc_compassion/tools/import_letter_functions.py [hunk 28]
```diff
@@ -21,13 +21,133 @@
...
+    temp_test = None
+    if (line_vals['child_id'] is None and line_vals['partner_id'] is None and
+            values['barcode'] == ""):
+        temp_test = True
+    # 2) we check that the partner and the child are the same for the computed
+    # and the expected values
+    elif (partner_id.id == line_vals['partner_id'] and
+          child_id.id == line_vals['child_id']):
+        temp_test = True
+    # 3) if 2 does not work, it means that something was wrong with the
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: So you don't have the scenario when only one information is missing (either child or sponsor) ?

File: sbc_compassion/tools/import_letter_functions.py [hunk 29]
```diff
@@ -21,13 +21,133 @@
...
+        error = error + 'template,\n'
+
+    # second: qr codexb
+    partner_id = env['res.partner'].search([
+        ('id', '=', line_vals['partner_id'])])
+    child_id = env['compassion.child'].search([
+        ('id', '=', line_vals['child_id'])])
+    # 1) we check the case where the barcode was not detected and was not
+    # expected
+    temp_test = None
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: You could define a function that decodes the barcode data into partner_ref and child_code values (can be used both in the import process and in this test). Here you could then do something like this, which would be more easy:
```
partner_csv, child_csv = decode_sponsorship_barcode(values['barcode'])
if partner_csv != partner.ref or child_csv != child.code:
    error = error + 'barcode,\n'
```

File: sbc_compassion/tools/import_letter_functions.py [hunk 30]
```diff
@@ -21,13 +21,133 @@
...
+    # Test that the template is the same as given in the csv
+    template_id = env.ref(
+        'sbc_compassion.'+values['template_id'])
+    tpl_detected = env['sponsorship.correspondence.template'].search(
+        [('id', '=', line_vals['template_id'])])
+    if template_id != tpl_detected:
+        error = error + 'template,\n'
+
+    # second: qr codexb
+    partner_id = env['res.partner'].search([
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: rename to `partner_detected`

File: sbc_compassion/tools/import_letter_functions.py [hunk 31]
```diff
@@ -21,13 +21,133 @@
...
+        ('id', '=', line_vals['partner_id'])])
+    child_id = env['compassion.child'].search([
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: `child_detected`

File: sbc_compassion/tools/import_letter_functions.py [hunk 32]
```diff
@@ -21,13 +21,133 @@
...
+    # 3) if 2 does not work, it means that something was wrong with the
+    # detection
+    else:
+        temp_test = False
+    if not temp_test:
+        error = error + 'barcode,\n'
+
+    # third: languages
+    temp_test = True
+    for lang in line_vals['test_letter_languages_id'].split(','):
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Since you use twice `line_vals['test_letter_languages_id'].split(',')`, you can put in in a variable before.

File: sbc_compassion/tools/import_letter_functions.py (hunk 28 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: If I have a partner missing, partner_id.id == False and line_vals['partner_id'] == False too, therefore the test will be OK

File: sbc_compassion/tools/import_letter_functions.py [hunk 33]
```diff
@@ -159,15 +274,42 @@ def _find_qrcode(env, img, MULTIPAGE, file_,
...
+def decodeBarcode(env, barcode):
+    """ Split the barcode and return the id of the partner and the child.
+    If the partner is not found, return None (same for the child).
+    :param env: Odoo environment
+    :param Barcode barcode:
+    :return: partner.id, Child.id
+    :rtype: int, int
+    """
+    if barcode is not None and 'XX' in barcode.data:
+        partner_id, child_id = barcode.data.split('XX')
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: If `barcode` contains more than one 'XX', split will return an array with len() > 2 and will break the code. You should test len(split()) == 2.
Also rename `partner_id`and `child_id` to `partner_ref` and `child_code` and later when you search you can use `partner_id` and `child_id`

File: sbc_compassion/tools/import_letter_functions.py [hunk 34]
```diff
@@ -159,15 +274,42 @@ def _find_qrcode(env, img, MULTIPAGE, file_,
...
+            'qr_preview': test_data
+        })
+    return dict_odoo, data, img
+
+
+def decodeBarcode(env, barcode):
+    """ Split the barcode and return the id of the partner and the child.
+    If the partner is not found, return None (same for the child).
+    :param env: Odoo environment
+    :param Barcode barcode:
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Why is barcode a Tuple and not a simple string ?

Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: It is not a Tuple, but a class (see zxing.Barcode).
In the test case, I am just using a namedtuple in order to mimic the behavior of the barcode

File: sbc_compassion/tests/test_tools.py [hunk 35]
```diff
@@ -60,8 +60,8 @@ def test_pattern_recognition(self):
         self.assertIsNotNone(template)
         self.assertEqual(template.name, 'Test Template 2')
         np.testing.assert_allclose(pattern_center,
-                                   np.array([1273.47, 3298.65]),
-                                   atol=0.1)
+                                   np.array([1273, 3301]),
```
Comment (ecino) [inline sbc_compassion/tests/test_tools.py]: @rz-comp can you validate this change ?

File: sbc_compassion/tests/test_tools.py [hunk 36]
```diff
@@ -121,7 +121,8 @@ def _read_img(self, path):
 
     def _pattern_recognition(self, path):
         img = self._read_img(path)
-        return tools.patternrecognition.find_template(img, self.templates, 0.3)
+        return tools.patternrecognition.find_template(
+            img, self.templates, 0.3)[:2]
```
Comment (ecino) [inline sbc_compassion/tests/test_tools.py]: I'm not fan of this, as it seems a bit "magic" to cut the result. By reading the code we don't understand what's going on. Can you return all data and just ignore what is not needed in the tests ?

File: sbc_compassion/tools/import_letter_functions.py [hunk 37]
```diff
@@ -21,13 +21,122 @@
...
+    """ Test each line and verify the data given by the csv.
+    If a line does not correspond to the csv, this function erases it.
+    A csv file need to be written in the same way than
+    sbc_compassion/tests/testdata/import/travis_file.csv
+
+    :param env: Odoo environment
+    :param dict line_vals: Dictionnary that will be used to create line
+    :param Recordset('ir.attachment') csv_file_ids: CSV used to check
+    :param str doc_name: Document name
+    :returns: test (True if equal to csv), error
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Update return doc

File: sbc_compassion/tests/test_tools.py (hunk 36 above)
Comment (loikki) [inline sbc_compassion/tests/test_tools.py]: If I am doing it, I will always ignore the last parameter (it will always be None).
Maybe it will be better if I am just explaining in a comment before returning the values

File: sbc_compassion/tools/import_letter_functions.py [hunk 38]
```diff
@@ -21,13 +21,122 @@
...
+        if csv_line['name'] == doc_name:
+            values = csv_line
+            break
+    else:
+        raise Warning(
+            _("File '{}' not found in the csv files".format(doc_name)))
+    # Test that the template is the same as given in the csv
+    template_csv = env.ref(
+        'sbc_compassion.'+values['template_id'])
+    tpl_detected = env['sponsorship.correspondence.template'].search(
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: I think you don't need to search for the template. You can simply do:
`if template_csv.id != line_vals['template_id']:`

File: sbc_compassion/tools/import_letter_functions.py [hunk 39]
```diff
@@ -21,13 +21,122 @@
...
+    if template_csv.id != tpl_detected.id:
+        error = error + 'template,\n'
+
+    # Test that the barcode detection matches the result given in csv file
+    BarcodeTuple = namedtuple('BarcodeTuple', ['data'])
+    barcode = BarcodeTuple(values['barcode'])
+    partner_detected, child_detected = decodeBarcode(env, barcode)
+    # 1) we check the case where the barcode was not detected and was not
+    # expected
+    if (line_vals['child_id'] != child_detected or
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Will `line_vals` always contain the `child_id` key ? If not, it should be replaced by `line_vals.get('child_id')` to avoid KeyErrors

File: sbc_compassion/tools/import_letter_functions.py [hunk 40]
```diff
@@ -21,13 +21,122 @@
...
+            line_vals['partner_id'] != partner_detected):
+        error = error + 'barcode,\n'
+
+    # third: languages
+    temp_test = True
+    letter_languages_id = line_vals['test_letter_languages_id'].split(',')
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: What if `line_vals['test_letter_languages_id']` is empty ? split will raise an error

File: sbc_compassion/tools/import_letter_functions.py (hunk 39 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: If line_vals comes from my analyze attachment (which is the case actually), I will always have theses keys

File: sbc_compassion/tools/import_letter_functions.py (hunk 40 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: See comments above about the fact that this value exists.
Split does not raise an error when the string is empty.

File: sbc_compassion/tools/import_letter_functions.py [hunk 41]
```diff
@@ -21,13 +21,122 @@
...
+    barcode = BarcodeTuple(values['barcode'])
+    partner_detected, child_detected = decodeBarcode(env, barcode)
+    # 1) we check the case where the barcode was not detected and was not
+    # expected
+    if (line_vals['child_id'] != child_detected or
+            line_vals['partner_id'] != partner_detected):
+        error = error + 'barcode,\n'
+
+    # third: languages
+    temp_test = True
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: rename to `lang_ok`

File: sbc_compassion/tools/import_letter_functions.py [hunk 42]
```diff
@@ -21,13 +21,122 @@
...
+    BarcodeTuple = namedtuple('BarcodeTuple', ['data'])
+    barcode = BarcodeTuple(values['barcode'])
+    partner_detected, child_detected = decodeBarcode(env, barcode)
+    # 1) we check the case where the barcode was not detected and was not
+    # expected
+    if (line_vals['child_id'] != child_detected or
+            line_vals['partner_id'] != partner_detected):
+        error = error + 'barcode,\n'
+
+    # third: languages
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Test that languages detected are the same as given in the csv file

File: sbc_compassion/tools/import_letter_functions.py [hunk 43]
```diff
@@ -21,13 +21,122 @@
...
+        temp_test = temp_test and lang in values['lang']
+    line_nber_lang = len(letter_languages_id)
+    csv_nber_lang = len(values['lang'].split(','))
+    temp_test = temp_test and line_nber_lang == csv_nber_lang
+
+    if not temp_test:
+        error = error + 'lang,\n'
+
+    # remove the two last character (',\n')
+    return error[:-2]
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: If there are no errors, this will fail

Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: No it will not

File: sbc_compassion/tools/import_letter_functions.py [hunk 44]
```diff
@@ -158,46 +273,94 @@ def _find_qrcode(env, img, MULTIPAGE, file_,
...
+    """
+    if barcode is not None and 'XX' in barcode.data:
+        barcode_split = barcode.data.split('XX')
+        if len(barcode_split) == 2:
+            partner_ref, child_code = barcode_split
+            child_id = env['compassion.child'].search(
+                [('code', '=', child_code)]).id
+            partner_id = env['res.partner'].search(
+                [('ref', '=', partner_ref),
+                 ('is_company', '=', False)]).id
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: You can add a condition, `contract_ids != False` to be sure only one partner will be found (sometimes we have duplicates in production (due to employees creating twice the same partner).

File: sbc_compassion/tools/import_letter_functions.py [hunk 45]
```diff
@@ -158,46 +273,94 @@ def _find_qrcode(env, img, MULTIPAGE, file_,
...
+        else:
+            partner_id = None
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: You can initialize partner_id and child_id to None at the beginning and remove those else parts

File: sbc_compassion/tests/test_tools.py (hunk 36 above)
Comment (ecino) [inline sbc_compassion/tests/test_tools.py]: If its always None then you need to update `find_template` method and remove that return.

File: sbc_compassion/tools/import_letter_functions.py (hunk 39 above)
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Fine

File: sbc_compassion/tools/import_letter_functions.py (hunk 43 above)
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Ok :wink:

File: sbc_compassion/tools/import_letter_functions.py (hunk 40 above)
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Fine

File: sbc_compassion/tools/import_letter_functions.py (hunk 44 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: Like this?
```python
partner_id = env['res.partner'].search(
    [('ref', '=', partner_ref),
    ('is_company', '=', False),
    ('contract_ids', '!=', False)]).id
```

File: sbc_compassion/tests/test_tools.py (hunk 36 above)
Comment (loikki) [inline sbc_compassion/tests/test_tools.py]: In this case, it is always None (you need to give the parameter test=True to obtain a non-None value)

File: sbc_compassion/tools/import_letter_functions.py (hunk 44 above)
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Sorry I think this won't work after checking. We could then check it has the "Sponsor" category but this is more complex than one search.

File: sbc_compassion/tests/test_tools.py (hunk 36 above)
Comment (ecino) [inline sbc_compassion/tests/test_tools.py]: Ok then add a comment.

File: sbc_compassion/tools/import_letter_functions.py [hunk 46]
```diff
@@ -21,13 +21,119 @@
 import checkboxreader as cbr
 import bluecornerfinder as bcf
 import numpy as np
+from math import ceil
 from glob import glob
 from wand.image import Image
+from collections import namedtuple
+
+from openerp import _
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: Add import exceptions.Warning

File: sbc_compassion/tools/import_letter_functions.py [hunk 47]
```diff
@@ -21,13 +21,119 @@
...
 
 
 ##########################################################################
 #                           GENERAL METHODS                              #
 ##########################################################################
+def testline(env, line_vals, csv_file_ids, doc_name):
+    """ Test each line and verify the data given by the csv.
+    A csv file need to be written in the same way than
+    sbc_compassion/tests/testdata/import/travis_file.csv
```
Comment (ecino) [inline sbc_compassion/tools/import_letter_functions.py]: No, it is not the same way. In travis_file, there is no 'barcode' column, but in this csv there should be one, otherwise the test fails.

**Comment by ecino** [comment]
Languages cut result should be shown even if no language is detected (if template is detected)

File: sbc_compassion/tools/import_letter_functions.py (hunk 47 above)
Comment (loikki) [inline sbc_compassion/tools/import_letter_functions.py]: I have corrected the csv in the test

File: sbc_compassion/tests/test_tools.py (hunk 35 above)
Comment (rz-comp) [inline sbc_compassion/tests/test_tools.py]: :+1:

**Comment by ecino** [comment]
:+1:
